include/boost/math/differentiation/autodiff.hpp

```diff
@@ -6,10 +6,13 @@
 #ifndef BOOST_MATH_DIFFERENTIATION_AUTODIFF_HPP
 #define BOOST_MATH_DIFFERENTIATION_AUTODIFF_HPP
 
-#include <boost/config.hpp>
+#include <boost/cstdfloat.hpp>
+#include <boost/lexical_cast.hpp>
```

Comment:
Is lexical_cast.hpp absolutely necessary here? I ask because if you run `#include <boost/lexical_cast.hpp>` you get an additional 410 header dependencies! (But I guess you already checked that `static_casts` won't work?)

Reply:
There are a number of instances, but to take one example, `boost::math::tgamma1pm1(x)` at some point calls `boost::lexical_cast<Real>(x)` where `Real=autodiff_fvar<>`. Adding an `explicit fvar(char const* ca)` constructor which calls `boost::lexical_cast(ca)` seemed to fix it, but perhaps there is a better way?

Reply:
Ok, if it's already being pulled in then there's no additional compilation cost.

Reply:
Sorry, I meant to get to this earlier today, but Matthew's answered the question.

For a little more color, I looked back at the PR which first included this constructor (https://github.com/pulver/autodiff/pull/11); Matthew asked me for the reason for the new constructor, and below is the explanation. 

"During static initialization, `unchecked_factorial` (actually `unchecked_factorial_imp`) calls `boost::math::tools::convert_from_string<T>(const char*)`, which returns either `const char*` or type `T` (depending on whether `T` is constructible from `const char*`). Without the constructor, we'll need a way to have `boost::lexical_cast` (called by `boost::math::tools::convert_from_string`) operate on the root type rather than on `autodiff_fvar` directly. The extra code in the constructor is due to `float` causing `boost::lexical_cast` to throw an error when the factorials are too large too fit within the `float` type.

I agree, though: it's not an ideal situation."

Reply:
Btw, what's the recommended way to define types `T` in order to call functions like `boost::math::tgamma1pm1(x)` which require the type `T` to work with `boost::lexical_cast<T>`?

We solved it by adding the `fvar(char const*)` constructor but am looking to remove this from the API.

CC: @pabristow @NAThompson @kedarbhat

Reply:
I just took a look at that and it looks like `std::to_string` will work instead of lexical cast in this case. I believe that that would be a backwards incompatible change from C++03, so @jzmaddock will have to give me the thumbs up to make this change:

```cpp
➜  special_functions git:(develop) grep -nr 'lexical_cast' .
./detail/unchecked_factorial.hpp:18:#include <boost/lexical_cast.hpp>
./detail/polygamma.hpp:147:       return policies::raise_evaluation_error<T>(function, ("Exceeded maximum series evaluations evaluating at n = " + boost::lexical_cast<std::string>(n) + " and x = %1%").c_str(), x, pol);
./lambert_w.hpp:187:{ // try a lexical_cast and hope for the best:
./lambert_w.hpp:188:   return boost::lexical_cast<double>(z);
```